### DISCUSSION TITLE
Questions about the security and license

### DISCUSSION THREAD

**Discussion by WangJiaoJo** [codesandbox/sandpack#960] [General]
Hi, team,

We have some security questions about the @codesandbox/sandpack library and hope to confirm with you.
We noticed that when the Sandpack rendering the code the code is rendering on the https://1-19-0.codesandbox.io/, because of the company security policy, our code is not allowed to expose to external. 
So our questions are
Will our code be uploaded to and saved at the CodeSandbox server?
If the answer is yes of the question 1, then how codesandbox can make sure our code not be exposed or prevent the hacking?
We use the CodeSandbox Sandpack to render the code in our web application, if we leave the page or close the application, will the uploaded code be destroyed or saved at CodeSandbox server?
Is there any solution or action we can take to make sure our code will be in the security if using the CodeSandbox Sandpack?
Or do we need to purchase the commerchial license for the code security? If yes, could you please provide the price?

Thanks in advance!

**Comment by danilowoz** [comment]
Hi! Sandpack doesn't store, save, or even upload any file to CodeSandbox server as long as you don't click on the "Open Sandbox" button. CodeSandbox domain only hosts the bundle files to execute your code and make sure the sandbox doesn't have access to the main domain cookies.

If this is still a concern, you can self-host the bundler by following these instructions: https://sandpack.codesandbox.io/docs/guides/hosting-the-bundler
